src/voices/voices.rs

```diff
@@ -53,44 +59,58 @@ where
     }
 }
 
-pub struct Voices {
-    voices: BTreeMap<WithCreationTime<Key>, Sink>,
+pub struct Voices<T> {
+    voices: BTreeMap<WithCreationTime<Key>, T>,
     capacity: usize,
 }
 
-impl Voices {
+impl <T> Voices<T> {
     pub fn new(capacity: usize) -> Self {
         Self {
             voices: BTreeMap::new(),
             capacity,
         }
     }
 
-    fn is_playing(&self, key: &Key) -> bool {
-        self.voices.contains_key(&WithCreationTime::new(*key))
+    fn is_playing(&self, key: Key) -> bool {
+        self.voices.contains_key(&WithCreationTime::new(key))
+    }
+
+    pub fn play(&mut self, key: Key, sink: T) -> Option<&T> {
+        if !self.is_playing(key) {
+            if self.voices.len() == self.capacity {
+                self.voices.pop_first();
+            }
+            let w = WithCreationTime::new(key);
+            self.voices.insert(w, sink);
+            self.voices.get(&w)
+        } else {
+            None
+        }
     }
 
-    pub fn play(&mut self, key: Key, sink: Sink) -> Option<&Sink> {
-        if !self.is_playing(&key) {
+    pub fn play_with_time(&mut self, key: Key, time: SystemTime, sink: T) -> Option<&T> {
+        if !self.is_playing(key) {
             if self.voices.len() == self.capacity {
                 self.voices.pop_first();
             }
-            self.voices.insert(WithCreationTime::new(key), sink);
-            self.voices.get(&WithCreationTime::new(key))
+            let w = WithCreationTime::new_with_time(key, time);
```

Comment:
I think the original code calling the constructor twice may be a bug? You may end up with different timestamps in each, and you want your map lookup to be based on the `Key` only I believe, not the time.